shaping: initial implementation of harfbuzz integration

whereswaldon opened the pull request:
This is the code I used to do shaping with our interfaces. I'm well aware that it can be improved, and I look forward to your suggestions on how to do so!

~~That being said, I'd _really_ prefer to turn `Input` and `Output` into structures. They represent data, not behavior, so an interface doesn't seem to be the right abstraction to me. I've left that change out of this PR though.~~ This PR now includes converting those interfaces into structures.

whereswaldon committed:
shaping: initial implementation of harfbuzz integration

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

stuartmscott commented on shaper.go:
```diff
@@ -1,6 +1,141 @@
+	buf := harfbuzz.NewBuffer()
+	runes, start, end := input.Text()
+	buf.AddRunes(runes, start, end-start)
+	// TODO: handle vertical text?
```
https://github.com/go-text/di/blob/master/direction.go#L9

whereswaldon replied:
Yup, just wanted to note that this is where it would be done.

stuartmscott commented on shaper.go:
```diff
@@ -1,6 +1,141 @@
+			if !ok {
+				// TODO: can this error happen? Will harfbuzz return a
+				// GID for a glyph that isn't in the font?
+				panic(fmt.Errorf("font missing glyphs for: %v", g.GlyphInfo.Glyph))
```
I wonder if Shape should return `(Output, error)` so toolkits can handle this rather than panicing

whereswaldon replied:
Fine by me.

stuartmscott commented on shaper.go:
```diff
@@ -1,6 +1,141 @@
+		tallest  int32
+	)
+
+	for i := range out.Glyphs {
+		g := &out.Glyphs[i]
+		switch input.Direction() {
+		case di.DirectionLTR, di.DirectionRTL:
```
Would it be more efficient to swap the nesting here, so `select` then `for`? Just feels wasteful to check something multiple times when we know it wont change.

whereswaldon replied:
Good idea!

stuartmscott commented on shaper.go:
```diff
@@ -1,6 +1,141 @@
+// TODO: can we just make Input and Output into structs?
+// Interfaces encapsulate varying behavior well, but these
+// types have no behavior. They're just structured data.
+// It feels silly to wrap them in accessors like this.
```
I see your point, it is rather verbose and they don't encapsulate any behaviour

stuartmscott commented on go.mod:
```diff
@@ -3,6 +3,7 @@ module github.com/go-text/shaping
 go 1.15
 
 require (
+	github.com/benoitkugler/textlayout v0.0.1
```
I presume this (MIT) is compatible with both toolkits' licenses (BSD-3-Clause and GNU Lesser General Public)

whereswaldon replied:
I know it's compatible with Gio's (MIT/UNLICENSE). I do not believe that there are conflicts with the other two you mention.

stuartmscott replied:
Oh silly me - I thought Gio was LGPL, but turns out I looked at the wrong wikipedia entry https://en.wikipedia.org/wiki/GIO_(software)

stuartmscott commented on shaper.go:
```diff
@@ -1,6 +1,141 @@
+	}
+	buf.Props.Language = input.Language()
+	buf.Props.Script = input.Script()
+	// TODO: figure out what (if anything) to do if this type assertion fails.
```
I see a couple of options here, we could either return an error, or make input.Face return `harfbuzz.Face` if that is the only thing we support, or we should move `Shape` into more specific packages, eg create a `harfbuzz` package with `Shape(Input) (Output, error)` where those `Input`/`Output` extend a base version, which could be accompanied by an `image` package also containing a `Shape` function but which uses `golang.org/x/image/font.Face`

whereswaldon replied:
As I do not currently anticipate us using anything other than a `harfbuzz.Face`, I'd be fine changing the type of `input.Face`. Your other option is viable, but seems like overkill. I'm not aware of a use-case for the image/font shaping if harfbuzz shaping is an option.

stuartmscott commented on shaper.go:
```diff
@@ -1,6 +1,141 @@
+	upem := sfnt.Units(font.Face().Upem())
+	ppem := input.Size()
+
+	// Actually use harfbuzz to shape the text.
```
Funny, I thought this project would eventually be its own shaper rather than just plumbing for another. Are you just using harfbuzz to get it working, or do you think that will be the shaper longer term?

whereswaldon replied:
We have opposite expectations. Since @benoitkugler has already gone through the trouble of porting harfbuzz to Go, I'd like to reuse that work rather than reimplementing it. In case it's the name that is causing confusion, this is leveraging the Go port, not the C++ codebase.

My perception of this project was to develop common code for accessing shaping as well as higher-level things like bidi splitting, paragraph layout, and font management.

stuartmscott replied:
Ack, makes sense

benoitkugler replied:
I would suggest that, at long term, we aim at moving the implementation from `github.com/benoitkugler/textlayout` to this repository. 
I think it would be valuable for two reasons :
- probably better code hygiene on a shared, "official" repository 
- the scope of the shaper, as it is now, is very close to (almost directly copied from) what harfbuzz is doing, so that having two packages for the same task feels a bit odd.

However, since it's not an easy task it seems fine to start by just importing `github.com/benoitkugler/textlayout` and using it. 
When we get a better understanding of the process, we could then think of merging.

benoitkugler replied:
In the same idea, concerning the `harfbuzz.Face` interface, notice that [this PR](https://github.com/go-text/font/pull/1) is a first step towards our own `Face`, with the long term goal to integrate font management into this project.

stuartmscott reviewed (commented):
Thanks for working on this @whereswaldon!

benoitkugler commented on shaper.go:
```diff
@@ -1,6 +1,141 @@
+// Shape turns an input into an output.
+func Shape(input Input) Output {
+	// Prepare to shape the text.
+	// TODO: maybe reuse these buffers for performance?
```
That is indeed a possible optimisation, since the `harfbuzz.Buffer` is designed to reuse allocated memory.
For instance, the Pango C library use a global variable to store one buffer (with a mutex) and reuse it.

whereswaldon replied:
This would be interesting to explore down the road. Thanks!

whereswaldon committed:
shaping: return error on missing glyphs

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: check direction outside of loop

Minor optimization to avoid rechecking text direction each
iteration.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: convert input and output to structs

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon commented:
Hey all, I've updated this PR to reflect the conversation so far:

- returns an error instead of panicking when a font is missing a needed glyph
- the direction of the text is now not checked in a loop
- Input and Output are now structs instead of interfaces

I'd appreciate you taking another look!

benoitkugler commented:
Thank you for the new changes, which seem fine to me !

Two comments/ideas (probably not urgent):

- the call to `harfbuzz.NewFont()` may be a bit expensive, since the layout tables (GPOS/GSUB) are processed and cached. Since Shape() is likely to be called many times (on small inputs) with the same font, it may be a good idea to cache the result of `harfbuzz.NewFont()`. We can either do it internally, but this requires that `font.Face` is a valid map key, or let the user do it, by requiring that `Input` has a `Font *harfbuzz.Font` field instead of `Face font.Face`.
- I think the scaling can be done directly by Harfbuzz (instead of using `scalePosition`), by using the `XScale` and `YScale` of the `harfbuzz.Font` struct. In our case `font.XScale, font.YScale = input.Size, input.Size` should work. I need to check that though, maybe when we have some basic tests.

benoitkugler commented on input.go:
```diff
@@ -1,24 +1,30 @@
-	// Face returns the font face to render the text in.
-	Face() font.Face
-	// Size returns the size of the font, eg. 14.
+type Input struct {
```
Nitpicking : maybe use a regular named field for `Direction`, `Face`, `Script`, `Language`, since there is no idea of type hierarchy (inheritance) here ?

stuartmscott reviewed (approved):
LGTM

(nit: you might want to update the last paragraph of the opening message 😛)

whereswaldon committed:
shaping: use harfbuzz to scale; fix bounds and baseline

This commit takes advantage of harfbuzz's builtin
scaling features to avoid converting from font units
manually. Additionally, it corrects several mistakes
related to computing the bounds and baseline of the
text.

This commit does not return the smallest possible bounding
box right now, but takes a more conservative approach
based upon the font's ascent and descent height.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
deps: update to textlayout v0.0.2

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: remove manual scaling

The updated textlayout dependency does this for us.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: compute accurate minimum bounding box

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: avoid embedding fields in input

The embedding wasn't useful, and could promote methods
from those fields to the entire type unexpectedly.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon commented:
Okay, I know I've had you look at this a million times, but I've been iterating on this and using it from Gio for a while now. The latest pushed changes are correct to the best of my understanding, and appear to produce correct results when used.

Please double-check the way that I'm computing the bounds, advances, and baseline. This area went through many changes, and while I think it is now correct, I'd appreciate an expert opinion.

If these changes pass muster, I'd like to merge this. I think (barring incorrect bounding box calculation) that it's ready.

whereswaldon commented:
@benoitkugler I did update to use the harfbuzz scaling, but I'd like to wait on updating the type of the Face. Thanks for those tips!

whereswaldon committed:
shaping: remove unused scaling funcs

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

stuartmscott commented on output.go:
```diff
@@ -1,19 +1,80 @@
+// the Output to match the current contents of the Glyphs field. This
+// operation requires querying extra information from the font as well
+// as knowing the direction of the text.
+func (o *Output) Recalculate(dir di.Direction, font *harfbuzz.Font) error {
```
Is this expected to be called by anyone other than the shaper? If not, exporting it pollutes the public API and it might be cleaner if the shaper calculates the values and set the fields so Output remains a 'dumb' struct

whereswaldon replied:
I could be doing this wrong, but I found that I needed this functionality during paragraph shaping. When I chose to line-break a run of text, I needed to recompute the width of the two new lines. That re-computation requires knowing the left and right bearings of the text on either end, which is why this functionality was necessary. @stuartmscott @benoitkugler Are you aware of a different way to determine the width of the broken line, such that I wouldn't need this? An alternative would be to include the extents as part of the `Glyph` type so that they didn't need to be looked up again.

benoitkugler replied:
Looking at the way Pango is working, it indeed seems that you need to compute the width of an "sub-line".
However, I found that Pango only rely on the width of each glyph, which is taken as the `XAdvance` of the harbuzz output.
Face.GlyphExtents() is only used to compute the extents of the whole line (or run) after line-breaking, but is not used during line-breaking.
I have to admit that I don't grasp the difference between what Pango calls line extents and line width...

Still, my conclusion is that

- Only XAdvance is needed to line break
- A method returning the width of a portion of the output is indeed needed by higher level packages

whereswaldon replied:
@benoitkugler Huh. I don't understand how pango's results can be visually accurate if it works like that. The left and right extents of the line can go significantly past the width described by the XAdvance. I'd expect using only XAdvance to line-break would yield significantly worse visual alignment of the edges of lines than taking the extents into account. But then again, I'm new to all of this. Perhaps that's just the norm, and I've never noticed. It's certainly a simpler algorithm that way.

Additionally, our API's `Output` type returns a `Bounds` that is defined as: `Bounds is the smallest rectangle capable of containing the shaped text.` I've been interpreting that visually, which means taking the extents into account. Is that how you understood the meaning of that field? That it's the dimensions of a pixel box large enough to contain all parts of all glyphs in the run?

benoitkugler replied:
I'm not sure I understand your comment about left and right extents and I think we have different naming conventions.

From what I have understood, for a simple LTR horizontal run, the advance, that is the Pos.XAdvance field, is always greater than the width (or horizontal extent), that is the GlyphExtents.Width field. 

See for instance this small example : https://gist.github.com/benoitkugler/7d6599e387b8625705da68ee10c59223

If we take https://freetype.org/freetype2/docs/glyphs/metrics.png as a reference, we have the following matching :

- advance <-> Pos.XAdvance
- width <-> GlyphExtents.Width
- bearingX <-> GlyphExtents.XBearing

The right side bearing is then deduced with : advance - width - bearingX

Thus, the advance already "includes" the side bearings. That would explain why Pango only need this information for line-break.

GlyphExtents is however, still needed to compute the Y extents, but I dont think it is necessary for line breaking, so, if we agree on the previous point, it would be interesting to provide the extents computation as a separate step.

What do you think ?

whereswaldon replied:
@benoitkugler Huh. I definitely understood the relationship between those values differently. Thank you for the example and for the visual reference. I looked for a diagram like that, but didn't find one. I think I need to go try to sanity-check myself using that reference and see how it impacts my results. Really, thank you! I don't have your domain expertise, so I won't be surprised at all if I'm just wrong about this. :D

whereswaldon replied:
> for a simple LTR horizontal run, the advance, that is the Pos.XAdvance field, is always greater than the width (or horizontal extent), that is the GlyphExtents.Width field.

@benoitkugler Well, I think part of the issue is that I'm working with _complex_ scripts for my test cases. My work has all been using Arabic to exercise more of the shaper features.

What I've learned from Arabic is:

- Width is the visual width of the character. It has no fixed relationship with the XAdvance, as many Arabic characters have an XAdvance of 0 while still having a width.
- I also often have negative XBearing, because the glyph visually begins "before" the current position of the dot. Such glyphs often also have a right side bearing extending past the Xadvance.

I agree with both the diagram you linked and your math, but I encounter instances in which the advance does _not_ account for the bearings fairly often.

I guess that's why I'm confused. Since (in complex scripts) the character can overflow both sides of the `xadvance` pretty easily, how can pango use only the `xadvance` to wrap lines? It seems that like that would lead to jagged edges with strokes that hung past the edge of a bounding box.

It's also possible that there's some typographical convention that you don't need to align strokes that fall beyond Xadvance at the paragraph level. Maybe they're considered inconsequential, or even designed to hang over the edge? I don't really know.

I'm also unclear on whether the Glyph's XOffset should be considered as part of the width math or not. Do you happen to have any guidance there?

benoitkugler replied:
I think the advance should still be enough. Here is why : quoting from Harfbuzz source, `Pos.XAdvance` is

>  How much the line advances **after** drawing this glyph when setting text in horizontal direction

and `Pos.XOffset` is  

> How much the glyph moves on the X-axis **before** drawing it, this should not affect how much the line advances.

So, the algorithm for drawing glyphs is the following (taken from pango) : 

1. Render the current glyph at (x + XOffset, y + YOffset), where (x, y) is the current point
2. Advance the current point to (x + XAdvance, y)
3. Render the next glyph...

Some glyphs may indeed have 0 as XAdvance, but I think that in this case, a previous glyph would already have incremented the advance, so that it would not be visually disturbing. 

Could you share an example of rune input (and font) you have tried ? I'm maybe over-simplifying here, I would like to check !

whereswaldon replied:
![image](https://user-images.githubusercontent.com/3922125/140083985-80790a90-a384-49ff-ac90-3c4722c2c598.png)

This visualization is of a word (الأحلام) in Arabic using [this font](https://fonts.google.com/specimen/Scheherazade+New?subset=arabic#standard-styles). The red-ish line beneath each glyph is the Xadvance of that glyph. Each glyph stair-steps 4 pixels downwards so that it's easier to see the boundaries between them.


The green line represents the width of each character. It's starting X position is `dot+g.XBearing`, and its ending X position is `dot+g.XBearing+g.Width`

Here's a scaled-up version to make seeing the lines easier:
![arabic-lines-scaled](https://user-images.githubusercontent.com/3922125/140085015-5f8d5ab5-b773-4b73-b66c-a475818691a9.png)

I hadn't actually added this visualization until this conversation, so thanks for asking. It's helpful to see these numbers in action.

In general, it looks like you're right. The only time that the bearings do crazy things is when two glyphs are designed to overlap. That _shouldn't_ happen at a line break :crossed_fingers: , so we can probably safely ignore the extents when both computing the width of a line and when doing line breaking. That will also considerably simplify things.

Given all of this though, I find that I'm confused about exactly when the extents actually do matter. I don't need them when rendering the glyphs. Are they just points of reference, or what?

benoitkugler replied:
Thanks a lot for the picture, very informative !

> Given all of this though, I find that I'm confused about exactly when the extents actually do matter. I don't need them when rendering the glyphs. Are they just points of reference, or what?

Good question ! Looking at the Pango code, it seems it serves two purposes : 

- Horizontal extents is used instead of Harbuzz advance as a fallback (it seems deprecated though)
- Pango provides high-level functions to query the extents of a line, a run or even a cluster. These functions are **not** used to decide where to break lines, but they could probably be used by the user, for instance when selecting text. I guess that extent is visually better than a box built on relying only on glyph advances ?

To conclude on this point, do you think it would make sense to expose Recalculate as a separate method, but not to call it when shaping ? So that the extra computation may be avoided when, say, only line breaking is needed.

whereswaldon replied:
Yeah, I think the sanest available choice is to surface a method that computes _just_ the advance width of an output. I find that I need to recompute it pretty often (after shaping) when doing line breaking, so making it a method on the output is still useful.

As for the extents, I agree that they aren't needed when line breaking, and I'm no longer convinced that the math I was doing with them is useful. However, I suspect they are needed when doing text selection. The tricky thing is, the only way to get accurate extent information is to re-configure the same font face at the same scale so that you can query for them in the same units as the rest of the dimensions.

Would it be reasonable to add the bearing/extent/dimension data to the `shaping.Glyph` type? Or do you think that it's too expensive to extract that data from the font to do it every time?

whereswaldon replied:
Actually, we still need to query the extents in order to provide a tight bounding box on the glyphs. I think, for now, that we should include the extents of each glyph in the Glyph type in order to avoid higher-level code needing to re-query for that information.

benoitkugler replied:
Alright !
I guess that it's OK for now not to worry too much on performances and handle it later if needed ?
On this point, I note that Pango caches the glyph extents. We can probably add a caching layer to the shaper if needed (something like a map[Face, size, Glyph]Extents)

andydotxyz replied:
> Is this expected to be called by anyone other than the shaper? If not, exporting it pollutes the public API and it might be cleaner if the shaper calculates the values and set the fields so Output remains a 'dumb' struct

Did we get to the bottom of this original question? If it might be useful in the future can we leave it unexported for now?

whereswaldon replied:
I think it may be useful to applications trying to get a tight bounding box on text after line breaking. Line breaking requires us to chop up the resulting output struct and recalculate parts of its dimensions, and I think providing helpers to do that is very reasonable. The RecalculateAdvance method is absolutely essential in the API for line breaking logic. We can leave the equivalent of this method unexported for now, but I expect that users will need it eventually.

andydotxyz replied:
I'm not sure I completely follow - can you illustrate how it is used perhaps?
It seems very similar to updating the `Input` and calling `Shape` again, to understand the difference may help.

whereswaldon replied:
You can break a line without reshaping if harfbuzz doesn't flag the boundary cluster as unsafe to break. This means that you can create a new line with something like:

```go
var out Output // assume we already have this from the shaper
newLine := out
newLine.Glyphs=newLine.Glyphs[start:end]
newLine.RecalculateAll()
```

andydotxyz replied:
OK, that is kindof cool. But if we don't have to re-shape then what does recalculate mean?
It seems like there is power in here, but perhaps we can find a clearer way to express this in a future item of work (can it be unexported for now?)

benoitkugler replied:
Harfbuzz only returns metrics for individual glyphs. The purpose of the `RecalculateAll()` method is to aggregate these metrics for the whole `Input` line (or run).
If we want to make this clearer, we could choose a "pure" approach : remove the line fields from Output (`Advance` and `GlyphBounds`), and make `RecalculateAll` return a struct containing them. 
(This will force the consumer to call it, but it will probably almost be the case anyway.)

andydotxyz replied:
Yes you might be right about that approach, it could clarify things.
I think I was getting confused because `Output` has become a pseudo-input, so the lines were blurred.
Of course I would not want to see more harfbuzz specifics influence our API design, so is this approach similar to how other font libraries would work as well?

Another approach could be that `Output` contains a method `func Subset(start, end int) Output` (terrible name, sorry) and that would return a new output representing the new calculation results. This way output is purely output and new calculations result in a new output so we avoid putting input parameters in an output struct.

stuartmscott reviewed (commented):
Still looks good to me, though would it be possible to add some test cases for the advances, bounds, etc so any changes would be detected?

benoitkugler reviewed (approved):
LGTM

andydotxyz commented:
Sorry I have not been able to test this in Fyne recently - a few back to back illnesses have kept me struggling with TODO list.
I hope to try it this week and give feedback.

whereswaldon committed:
shaping: add simple tests for Output.Recalculate

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: add tests for glyphs with wide extents

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: add test for glyph hanging below baseline

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: test bounds calculation for offset glyphs

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: err in Recalculate on unsupported direction

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: add descriptive errors to Shape

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: add dual MIT/Unlicense to code

These licenses should be compatible with essentialy all
projects, and are the standard licenses for Gio code.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon commented:
Okay all, I've added tests to lock in the behavior of `Output.Recalculate` and added some descriptive errors we can return under varying conditions. I've also taken the liberty of adding a maximally-permissive license that should be compatible with all projects. If there's a problem with this licensing, please let me know.

I believe this is ready for merge, and I'll do that in 24 hours if nobody objects.

Cheers,
Chris

stuartmscott commented:
Sorry, I've been on vacation so might have missed it, was there a discussion about licenses?

whereswaldon commented:
@stuartmscott No, there was not. As the author of this PR, I place my work into the public domain, and am willing to additionally apply other licenses. I pushed a license file that would place this code under both the MIT and Unlicense, which should be compatible with every imaginable other piece of software. I can't see how that license choice could cause problems for anyone, but I'm willing to hear other viewpoints. Do you have an alternative proposal?

stuartmscott commented:
Yup, makes sense that all of these libraries should have a license. I typically use Apache, Fyne uses BSD, and Gio uses MIT. I don't have a problem with MIT, just wondered how the decision was made and whether there should be a discussion to hear viewpoints before one is chosen.

andydotxyz commented:
Thanks for doing this and apologies for delays at my end, I am starting to test this now.
The first thing I noticed is that the glyph elements of the API are tied to harfbuzz - is this something we want to do? Won't it then be very difficult if we need to have different implementations?

andydotxyz commented on shaper.go:
```diff
@@ -1,6 +1,82 @@
+	buf.Props.Language = input.Language
+	buf.Props.Script = input.Script
+	// TODO: figure out what (if anything) to do if this type assertion fails.
+	font := harfbuzz.NewFont(input.Face.(harfbuzz.Face))
```
This crashes when doing first tests with Fyne, because we have glyph substitution so the font face is not a `harfbuzz.Face`.

benoitkugler replied:
Interesting ! Could you elaborate on "glyph substitution" (or point me to the relevant code) ?

We have not made progress on `font.Face`, but I think we have to use our own interface, to provide the font metrics needed by harbuzz.

andydotxyz replied:
As discussed on other channels, there is probably a larger discussion here about this API and usage of harfbuzz, should we:

a) Make it harfbuzz specific and move this up to the field signature (fixing the TODO)
b) Keep it implementation agnostic and remove harfbuzz from the public API of this PR
c) Commit this PR with harfbuzz in the exposed API, with an ambition to make it abstract one the go-text project has suitable abstraction over Font/Face.

In discussion it sounds like c) is desirable, but we should capture that conversation here and possibly add issues for any additional work required.

whereswaldon replied:
I think c is preferable. I'd prefer to wait until we have a suitable abstraction across different kinds of faces before hiding harfbuzz behind it, and I don't want this PR to languish in review limbo until that happens.

andydotxyz replied:
I agree that this is probably the road to take.
Normally I would prefer that we commit the designed abstraction, then implement it - but given the work that has already taken place I think it might make sense to work the other way.

andydotxyz commented:
OK, I have this hooked up and it's giving sane numbers, but I am struggling with one thing.
Is there a way to tell how tall the font is? in our old font library it was `Face.Metrics().Height` but I can't figure it out with harfbuzz.
I guess this could be the same as Baseline + Descent, but I want to check this without the string passed necessarily having full height or full descent.

whereswaldon commented:
@andydotxyz Use [`ExtentsForDirection`](https://pkg.go.dev/github.com/benoitkugler/textlayout@v0.0.1/harfbuzz#Font.ExtentsForDirection) to get one of these: https://pkg.go.dev/github.com/benoitkugler/textlayout@v0.0.1/fonts#FontExtents

andydotxyz commented:
Thanks @whereswaldon - I found the data through the Harfbuzz code.
Unfortunately this is in a different (unscaled?) coordinate system.
Would it be possible to add the font height info to this API do you think? (For example `Output.LineHeight`), that way I think the Fyne usage of this go-text shaper could complete.

benoitkugler commented:
> Unfortunately this is in a different (unscaled?) coordinate system.

@andydotxyz Indeed the result depends on the `XScale` and `YScale` of the `harfbuzz.Font `object. 
Even if the value may be found using the harfbuzz API, it indeed makes probably sense to expose it as a shortcut, since for now the `harfbuzz.Font` object is not exposed.

andydotxyz commented:
Thanks @benoitkugler. Seeing such large numbers internally I can understand why it is possible to use just int within harfbuzz.
However as this scaling happens internally then I think it will not be possible (or sensible) to try and use just int in our public API.

whereswaldon commented:
I agree that we should surface line height in the output. I'll do that in the next revision.

whereswaldon committed:
shaping: add font line dimensions to Output

This makes it possible for layout code to choose an appropriate line
height for text regardless of the contents of the text.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon commented:
I pushed a change that surfaces line dimensions. Just a height was insufficient, as it didn't tell you where to position the glyphs within the line. I decided to just surface all three values that the font provides.

andydotxyz commented:
> I pushed a change that surfaces line dimensions. Just a height was insufficient, as it didn't tell you where to position the glyphs within the line. I decided to just surface all three values that the font provides.

Thanks, I got the value I needed using `out.LineBounds.MaxAscent - out.LineBounds.MaxDescent` :).
I still have some issues where length from this does not match the rendered text - will look into this just now.

andydotxyz commented:
I have found one strange issue, if I use `out.Baseline` then mixed faces (regular vs italic) don't line up correctly.
However if I moved to use `out.LineBounds.MaxAscent` as baseline then it works correctly.
Is this expected?

whereswaldon commented:
@andydotxyz So the Baseline value in the output tells you where the baseline of the text tightly bounded by `Bounds` is. That's a small rectangle containing just the shaped glyphs. The `LineBounds` describes a large rectangle capable of holding any glyphs within the whole font. So when laying out a paragraph of text, it is likely correct to use out.LineBounds.MaxAscent to set the baseline if you're using the sum of the LineBounds attributes as the line height. However, an application may want to work purely in terms of the tighter bounding rectangle described by `Bounds` situationally.

I can see how the current naming/structure is confusing though. I'm open to renaming/organizing fields to make the difference clearer. Any suggestions?

andydotxyz commented:
Hmm, good points. With regards to the "outer" bounds we have `LineBounds`, that encompasses the required information. for the "inner" bounds the bound and baseline are at the top level of the struct. Would it be worth exploring using the same extents type for both?

whereswaldon commented:
Hmm. So the "width" of the line is the same whether you're looking at the glyph or line bounds. Perhaps we have:

```go
type Bounds struct {
    Ascent, Descent, Baseline fixed.Int_26_6
}

type Output struct {
    // other fields elided
    GlyphBounds Bounds
    LineBounds Bounds
    Advance fixed.Int_26_6
}
```
They share an "advance" value, but the minor axis dimensions are expressed in a consistent way? We'd be dropping the `Bounds` rectangle field and the root `Baseline` field.

whereswaldon committed:
shaping: temporarily include extents in Glyph type

This makes it easier to visualize the glyphs for debugging purposes.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: drop debug logging

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: use consistent type for bounds

This commit introduces a shared type that is used to both
describing the tight bounds of a set of glyphs and the
larger bounds of a 'line' in the current font.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon committed:
shaping: resolve extents in shaping

This commit simplifies output management by resolving the
font dimensions and glyph extents during shaping. The output
can have its Glyphs resliced and its bounds recomputed without
needing further access to the font data.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon commented:
I've just pushed some changes that do the following:

- Unified the bounding boxes so that they're all expressed in terms of a single `Bounds` type. This is used to express the tight glyph bounding box and the font's line bounding box.
- Updated the logic responsible for calculating the Advance field so that it now longer tries to take the right and left extents into account (@benoitkugler has me convinced that they should not be used for that).
- Extended the Glyph type to contain the glyph's extents information. This makes sense because:
    1. We already need to query the extent information for each glyph in order to compute the tight bounding box. We're already doing the work to compute the extents, so why not save it?
    2. Other code can't easily get the extent information later without reconstructing a harfbuzz.Face that is scaled identically to the one we use. It's painful and error-prone to expect other code to grab them later.
    3. It seems likely that many consumers of the shaping data will want the extents anyway, for purposes like correctly highlighting selected glyphs in a region of text.
    4. As a side-effect, this made the operations on `Output` into fairly pure functions, which _feels_ better than the previous API.

andydotxyz commented:
I have managed to adapt the Fyne code to use this. However occasionally the letter spacing seems strange. I wonder if it is possible that this line in my code is wrong? For each character painted we advance, but the `XAdvance` is an `int` which concerns me:

```go
	d.Dot.X += float32ToFixed266(float32(out.Glyphs[i].XAdvance) * canvasScale)
```

It can result in the following sort of strangeness:

![strange](https://user-images.githubusercontent.com/294436/140742632-248ef4e7-69c6-45ee-ae64-c10267b9ce34.png)

Should look more like:

![compare](https://user-images.githubusercontent.com/294436/140743026-5081e8f7-00d2-4130-8937-12558c555ad8.png)

whereswaldon commented:
@andydotxyz I believe that there are two possible issues:

1. How are you feeding the glyph data to the rasterizer? Does it take the `XOffset` and `YOffset` into account when you do it?
2. This may very well be a lack of pixel precision, in which case you'll need to scale up the precision at which you invoke harfbuzz and then scale it back after. I think this was discussed [here](https://github.com/go-text/shaping/issues/3#issuecomment-918452655). Given that harfbuzz only works in an integer space, I believe this is the intended way to achieve higher precision.

andydotxyz commented:
Confirmed that multiplying the harfbuzz font input up, doing shaping, then scaling back down again does fix the issue.
However I don't think we would be doing a good job if we required that API users do that. Particularly if we have fixed.Int26_6 in most parts of the API.

See in this image, multiplied up on left, without on right:
![font-comparison](https://user-images.githubusercontent.com/294436/140929599-d5eefd6d-2128-4f1f-b5af-e60bf83ba087.png)

whereswaldon commented:
Okay, so @andydotxyz , would the following address the scaling problem adequately for you?

Add a `Resolution int` field to the Input. This is the artificial scale factor applied to the input and divided out at the end. A value of 0 will be silently treated as a value of 1, and will result in the current effectively-integer output.

Change the `Glyph` type to be:

```go
type Glyph struct
    Width,Height,XBearing,YBearing,XAdvance,YAdvance,XOffset,YOffset fixed.Int26_6
    Cluster int
    Glyph fonts.GID
    Mask harfbuzz.GlyphMask
}
```

andydotxyz commented:
> Add a `Resolution int` field to the Input. This is the artificial scale factor applied to the input and divided out at the end. A value of 0 will be silently treated as a value of 1, and will result in the current effectively-integer output.

I guess I don't understand the value of exposing a workaround. If the output is `fixed.int26_6` should we not output those numbers, rather than always an `int` that looks like `Int26_6`?
What I mean is, what's the value in truncating by default?

benoitkugler commented:
> I guess I don't understand the value of exposing a workaround. If the output is `fixed.int26_6` should we not output those numbers, rather than always an `int` that looks like `Int26_6`?
> What I mean is, what's the value in truncating by default?

I think Chris has proposed this solution to use the internal scaling provided by Harfbuzz, which is convenient but only accepts integers. So there are two solutions:

- either use the resolution mechanism proposed by Chris. With a well chosen default resolution, we could probably make this setting optional
- either provide a constant scale to Harfbuzz and scale the result ourselves, which was what Chris initially wrote (At the time I felt it was a shame not to use the Harfbuzz mechanism.)

whereswaldon commented:
> I guess I don't understand the value of exposing a workaround. If the output is fixed.int26_6 should we not output those numbers, rather than always an int that looks like Int26_6?

A "workaround"? I don't understand. My intention was to give users a knob for the level of precision they want in the output. If you set it to 10, it means that your output `fixed.Int26_6` will have a fractional component with a precision of 1/10 (well, the closest that binary fractions allow anyway).

The reason that I approached it this way is because I don't know how to choose a reasonable default level of precision. How much is really needed? I also wanted to provide a way to opt out of the extra math that this does for every character, hence the 0 and 1 behavior.

I would be just as happy to make the output API integer-only. :D 

> What I mean is, what's the value in truncating by default?

I think maybe we're talking past one another. There is no truncation occurring at all. Harfbuzz works in terms of a specific ppem that we provide up front, and it returns integer measurements within that ppem. If you want higher-precision measurements, you have to artificially scale that ppem before providing it to harfbuzz, and then divide it out (in a non-truncating way) after. But you still need some way to decide the magnitude of the artificial scaling factor. Additionally, not all applications may need this scaling.

> I think Chris has proposed this solution to use the internal scaling provided by Harfbuzz

Yeah, I was thinking we'd mix this scaling factor into the math there, which makes the implementation simple (or at least I think it does).

andydotxyz commented:
> A "workaround"? I don't understand. My intention was to give users a knob for the level of precision they want in the output.

Maybe a poor choice of words, sorry - but as a developer using an API that shows `fixed.Int26_6` I would be surprised if they are always `int` values unless I specify some optional "precision" factor.

I agree I may not be on the same page as I have not worked with harfbuzz directly.
All I can do is look at this as an external developer. The inputs have a fractional component, but the output values we are talking about will, by default, not. So then the proposed scaling updates this behaviour to use franctional components, when specified a non-zero/one value.

> either use the resolution mechanism proposed by Chris. With a well chosen default resolution, we could probably make this setting optional

Yes I suppose it could make sense if the default is high enough to provide reasonable precision in standard usage.
But if we are taking this approach I come back to the question of why not just use fractional components as standard. If some usages really need to use int position for some reason (speed maybe?) then that could be discussed too.

stuartmscott commented:
I don't see why resolution/precision needs to be exported, it can be set internally by this library to the maximum supported/representable by `fixed.Int26_6`, so it scales everything up for harfbuzz and then scales it back down afterward - I don't see a use case where Gio or Fyne would want less than the maximum precision so it need not be in the public api.

khaledhosny commented:
FreeType normally uses 26.6 fixed-point format, so that might be a reasonable default.

whereswaldon commented:
I'll try setting the scaling factor to 64 internally, as I believe that will give us the maximum expressible precision in a `fixed.Int26_6`. We'll see how it comes out.

andydotxyz commented:
> I'll try setting the scaling factor to 64 internally, as I believe that will give us the maximum expressible precision in a `fixed.Int26_6`. We'll see how it comes out.

Sounds sensible, that should work well.

andydotxyz commented:
Let me know if there is anything I can do to help get this over the line.
One thing worth noting might be that the decision team have access to shared branches so we could potentially collaborate more directly on future work if it's a branch before becoming a PR.

whereswaldon committed:
shaping: use fixed-point with internal scaling

This commit converts the entire external API to use
fixed-point numbers that are automatically scaled to
their maximum resolution.

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon commented:
@andydotxyz I believe you already have push access to the base branch of this PR.

I've pushed an API update that converts everything to use fixed-point values with maximum scaling.

I'd like to propose that this PR be merged in its current state. I'm open to future API modifications, but I think the current API should at least be usable by everyone.

benoitkugler committed:
precise documentation about Size input

benoitkugler commented:
I've added some documentation about the `Size` field. Feel free to adjust the wording !

Nitpicking : Note that, so far, it may not only be a font size (aka point size, like 14), but anything the caller wants. So it would maybe be clearer to rename the field to `Scale` ?

Anyway this PR looks really clean, great job and thanks a lot Chris !

stuartmscott reviewed (commented):
Looks good!

I guess it is blocked on https://github.com/go-text/shaping/issues/6

andydotxyz reviewed (commented):
Thanks, this looks beautiful at non-integer scales by default now - thanks :)

I am OK to land the code as-is like you suggest, however with #6 still an open discussion I don't think that this PR should land with the LICENSE changes.

andydotxyz commented:
It sounds like there is a decision on #6.
Happy to update my review to approved once the license here matches, or is moved to a different PR.

whereswaldon committed:
license: relicense to BSD or Unlicense

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

whereswaldon commented:
@andydotxyz I've updated the license and SPDX metadata in this PR. If you approve, go ahead and merge.

andydotxyz reviewed (approved):
Thanks for all the hard work on this.
If possible future projects should probably land LICENSE separately to code when it's not the first commit. But we have discussed this enough to be comfortable this time I think.

eliasnaur commented on LICENSE:
```diff
@@ -1,34 +1,24 @@
 Both licenses are reproduced below.
 
 ----
-The MIT License (MIT)
```
Probably best to retain a title, to make it clear that the text below is the BSD-3 license.

whereswaldon replied:
Thanks, done!

whereswaldon committed:
license: add title above bsd text

Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>
